src/app/gibbs/single_thread_sampler.cpp

```diff
@@ -201,57 +111,76 @@ void SingleThreadSampler::sample_single_variable(long vid, bool is_inc) {
     }
   } else {
-    if (variable.domain_type == DTYPE_BOOLEAN) {
-      if (variable.is_evid == false || sample_evidence) {
-        potential_pos = p_fg->template potential<false>(variable, 1);
-        potential_neg = p_fg->template potential<false>(variable, 0);
+    if (variable.is_evid == false || sample_evidence) {
+      int proposal = draw_sample(variable, sample_evidence);
+      p_fg->template update<false>(variable, (double)proposal);
+    }
+  }
+}
+}
 
-        *this->p_rand_obj_buf = erand48(this->p_rand_seed);
-        if ((*this->p_rand_obj_buf) *
-                (1.0 + exp(potential_neg - potential_pos)) <
-            1.0) {
-          p_fg->template update<false>(variable, 1.0);
-        } else {
-          p_fg->template update<false>(variable, 0.0);
-        }
-      }
+int dd::SingleThreadSampler::draw_sample(Variable &variable,
+                                         bool is_free_sample) {
+  if (variable.is_evid && !is_free_sample) return variable.assignment_evid;
 
-    } else if (variable.domain_type == DTYPE_MULTINOMIAL) {
-      while (variable.upper_bound >= varlen_potential_buffer.size()) {
-        varlen_potential_buffer.push_back(0.0);
-      }
+  int proposal = 0;
 
-      if (variable.is_evid == false || sample_evidence) {
-        sum = -100000.0;
-        acc = 0.0;
-        multi_proposal = -1;
-        for (int propose = variable.lower_bound;
-             propose <= variable.upper_bound; propose++) {
-          varlen_potential_buffer[propose] =
-              p_fg->template potential<false>(variable, propose);
-          sum = logadd(sum, varlen_potential_buffer[propose]);
-        }
+  if (variable.domain_type == DTYPE_BOOLEAN) {
+    if (is_free_sample) {
+      potential_pos = p_fg->template potential<true>(variable, 1);
+      potential_neg = p_fg->template potential<true>(variable, 0);
+    } else {
+      potential_pos = p_fg->template potential<false>(variable, 1);
+      potential_neg = p_fg->template potential<false>(variable, 0);
+    }
 
-        *this->p_rand_obj_buf = erand48(this->p_rand_seed);
-        for (int propose = variable.lower_bound;
-             propose <= variable.upper_bound; propose++) {
-          acc += exp(varlen_potential_buffer[propose] - sum);
-          if (*this->p_rand_obj_buf <= acc) {
-            multi_proposal = propose;
-            break;
-          }
-        }
-        assert(multi_proposal != -1);
-        p_fg->template update<false>(variable, multi_proposal);
+    *this->p_rand_obj_buf = erand48(this->p_rand_seed);
+    // sample the variable
+    // flip a coin with probability
+    // (exp(potential_pos) + exp(potential_neg)) / exp(potential_neg)
+    // = exp(potential_pos - potential_neg) + 1
+    if ((*this->p_rand_obj_buf) * (1.0 + exp(potential_neg - potential_pos)) <
+        1.0) {
+      proposal = 1;
+    } else {
+      proposal = 0;
+    }
+  } else if (variable.domain_type == DTYPE_MULTINOMIAL) {
```

Comment:
Referring to the else block probably after this?

How about switching to `switch`, `case`s while we're at this?

We can probably just crash here (`abort()`?) on invalid `domain_type`, assuming the code that loads the variables print meaningful errors and prevents such cases here.